Internal/svTrainNN.ecl

```diff
@@ -0,0 +1,494 @@
+        * Then back-propagate errors (i.e. score - label)
+        * back to the output weights and the input weights.
+        */
+      void fbPropagate(tp& train)
+      {
+        //assert(train.main > 0 & train.main <= wc_);
+        //if (train.main < 1 || train.main > wc_)
+        //  return;
+        uint32_t main = train.main - 1; // Map 1-based ECL index to 0-based
+        uint32_t contxCount = train.contxCount;
+        uint32_t * contx = train.contx;
+        double contxCountInv = 1.0 / (double)contxCount;
+        // FORWARD PATH ***
+        uint32_t j, i, n, windx, target, label;
+        double cum, score, alpha, grad, wUpdate;
+        // Zero the hidden layer.
+        for (uint32_t i = 0; i < dim_; i++)
+          hidden_[i] = 0;
+        // Update the hidden layer outputs based the context words.
+        for (j = 0; j < contxCount; j++)
+        {
+          uint32_t cx = contx[j] - 1; // Map 1-based ECL index to 0-based
+          for ( i = 0; i < dim_; i++)
+          {
+              windx = toFlatIndx(0, cx, i);
+              hidden_[i] += w_[windx] * contxCountInv;
+          }
+        }
+        // Accumulate the final layer output for the 
+        // 'main' output node as well as a random selection of
+        // 'negative' words.
+        for (n = 0; n <= negSamp_; n++) // Note executes negSamp + 1 times
+        {
+          // On the first iteration, process the positive sample given
+          if (n == 0)
+          {
+            target = main;
+            label = 1;
+          }
+          else
+          // On subsequent iterations, process randomly selected negative
+          // samples.
+          {
+            target = getNegative(main);
+            label = 0;
+          }
+          cum = 0;
+          // Propagate the hidden layer to the output layer
+          for (j = 0; j < dim_; j++)
+          {
+            windx = toFlatIndx(1, j, target);
+            cum += w_[windx] * hidden_[j];
+          }
+          // Apply the sigmoid to the accumulated (weight * hidden)
+          // to get the final output value.
+          double score = sigmoid(cum);
+
+          // Calculate the loss, and add it to the total loss.
+          loss_ += calcLoss(score, label);
+
+          // BACK PATH ***
+          // Compute the gradient at the output layer
+          // = learning rate * (label - score) * derivative(lossFunc)
+          // derivative(lossFunc) = score * (1-score)
+          double grad2 = (float(label) - score) * score * (1-score);
+          // Apply the gradient to the output weights
+          // and the input weights.
+          for (j = 0; j < dim_; j++)
+          {
+            windx = toFlatIndx(1, j, target);
+            // Update output (i.e. L2) weights
+            // weight update = grad * hidden * weight
+            wUpdate = (hidden_[j] * grad2 * lr_);
+            w_[windx] += wUpdate;
+            wUpdates_[windx] += wUpdate;
```

Comment:
Rather than updating weights and weight updates, I suspect it would be more efficient to save the old weights, and then calculate the updates afterwards.  (Assuming each weight is updated more than a couple of times.)  As well as reducing the number of updates I suspect it would improve cache locality.